File: app/src/main/java/com/openclassrooms/realestatemanager/data/local/dao/SinglePropertyDao.java
package com.openclassrooms.realestatemanager.data.local.dao;

import android.database.Cursor;

import androidx.lifecycle.LiveData;
import androidx.room.Dao;
import androidx.room.Insert;
import androidx.room.OnConflictStrategy;
import androidx.room.Query;
import androidx.room.RawQuery;
import androidx.room.Transaction;
import androidx.room.Update;
import androidx.sqlite.db.SupportSQLiteQuery;

import com.openclassrooms.realestatemanager.data.local.entities.PropertyWithImages;
import com.openclassrooms.realestatemanager.data.local.entities.SingleProperty;

import java.util.List;

@Dao
public interface SinglePropertyDao {

    // change pour long
    @Insert(onConflict = OnConflictStrategy.REPLACE)
    long createSingleProperty(SingleProperty singleProperty);

    @Update
    int updateProperty(SingleProperty singleProperty);

    @Query("SELECT * FROM property WHERE pid = :propertyId")
    LiveData<SingleProperty> getSingleProperty(String propertyId);

    @Query("SELECT * FROM property WHERE pid = :propertyId")
    Cursor getSinglePropertyProvider(String propertyId);

    // What that return ?
//    @Query("SELECT pid,type,price FROM property")
//    public LiveData<List<SingleProperty>> getAllProperties();

    // get two table unified
    @Transaction
    @Query("SELECT * FROM property")
    LiveData<List<PropertyWithImages>> getPropertyWithImages();

    @Transaction
    @Query("SELECT * FROM property")
    Cursor getPropertyWithImagesProvider();

    @RawQuery
    LiveData<List<PropertyWithImages>> getPropertyWithImageQuery(SupportSQLiteQuery query);
}

File: app/src/main/java/com/openclassrooms/realestatemanager/data/local/dao/SinglePropertyDao.kt
package com.openclassrooms.realestatemanager.data.local.dao

import android.database.Cursor
import com.openclassrooms.realestatemanager.data.local.entities.SingleProperty
import androidx.lifecycle.LiveData
import androidx.room.*
import com.openclassrooms.realestatemanager.data.local.entities.PropertyWithImages
import androidx.sqlite.db.SupportSQLiteQuery

@Dao
interface SinglePropertyDao {
    // change pour long
    @Insert(onConflict = OnConflictStrategy.REPLACE)
    fun createSingleProperty(singleProperty: SingleProperty?): Long

    @Update
    fun updateProperty(singleProperty: SingleProperty?): Int

    @Query("SELECT * FROM property WHERE pid = :propertyId")
    fun getSingleProperty(propertyId: String?): LiveData<SingleProperty?>?

    @Query("SELECT * FROM property WHERE pid = :propertyId")
    fun getSinglePropertyProvider(propertyId: String?): Cursor?

    // get two table unified
    @get:Query("SELECT * FROM property")
    @get:Transaction
    val propertyWithImages: LiveData<List<PropertyWithImages?>?>?

    @get:Query("SELECT * FROM property")
    @get:Transaction
    val propertyWithImagesProvider: Cursor?

    @RawQuery
    fun getPropertyWithImageQuery(query: SupportSQLiteQuery?): List<PropertyWithImages?>?
}